src/garage/trainer.py

```diff
@@ -145,19 +145,19 @@ def setup(self, algo, env):
             policy weights can be loaded before setup().
 
         Args:
-            algo (RLAlgorithm): An algorithm instance.
+            algo (RLAlgorithm): An algorithm instance. If this algo want to use
+                samplers, it should have a `_sampler` field.
             env (Environment): An environment instance.
 
-        Raises:
-            ValueError: If sampler_cls is passed and the algorithm doesn't
-                contain a `max_episode_length` field.
-
         """
         self._algo = algo
         self._env = env
 
         self._seed = get_seed()
-        self._sampler = self._algo.sampler
+
+        if hasattr(self._algo, '_sampler'):
+            # pylint: disable=protected-access
+            self._sampler = self._algo._sampler
```

Comment:
Oh, I see. The idea here was always for algorithms to call `self._sampler` directly, instead of going through the trainer, but clearly that needs more work. This will do for now, I suppose.